web/libs/editor/src/lib/AudioUltra/Visual/Renderer/WaveformRenderer.ts

```diff
@@ -0,0 +1,260 @@
+    const renderId = this.config.renderId;
+    let x = 0;
+    const audio = this.audio;
+    const channelCount = audio?.channelCount ?? 1;
+    const height = this.config.waveHeight / channelCount;
+    const scrollLeftPx = context.scrollLeftPx;
+    const dataLength = context.dataLength;
+    this.lastWaveformRenderedScrollLeftPx = scrollLeftPx;
+    const samplesPerPx = context.samplesPerPx;
+    const shiftAmount = -deltaX;
+    const sampleDiff = Math.round(deltaX * samplesPerPx);
+    layer.shift(shiftAmount, 0);
+    for (let channelNumber = 0; channelNumber < channelCount; channelNumber++) {
+      let sStart = iStart;
+      let sEnd = iEnd;
+      if (deltaX > 0) {
+        sStart = iEnd - sampleDiff;
+        x = clamp(context.width + shiftAmount - 2, 0, context.width);
+      } else {
+        sEnd = iStart - sampleDiff;
+        x = 0;
+      }
+      sEnd = clamp(sEnd + samplesPerPx * 2, 0, dataLength);
+      const renderIterator = this.renderSlice(layer, height, sStart, sEnd, channelNumber, x, context);
+      const render = () => {
+        if (this.config.renderId !== renderId) return;
+        const next = renderIterator.next();
+        if (!next.done) {
+          requestAnimationFrame(render);
+        } else {
+          if (context && context.notifyRenderComplete) context.notifyRenderComplete();
+        }
+      };
+      render();
+    }
+  }
+
+  *renderSlice(
+    layer: Layer,
+    height: number,
+    iStart: number,
+    iEnd: number,
+    channelNumber: number,
+    x = 0,
+    context: RenderContext,
+  ): Generator<any, void, any> {
+    const audio = this.audio;
+    const bufferChunks = audio?.chunks?.[channelNumber];
+    if (!bufferChunks) return;
+    const bufferChunkSize = bufferChunks.length;
+    const paddingTop = this.config.padding?.top ?? 0;
+    const paddingLeft = this.config.padding?.left ?? 0;
+    const zero = height * channelNumber + (this.config.reservedSpace ?? 0);
+    const y = zero + paddingTop + height / 2;
+    let total = 0;
+    layer.save();
+    const waveColor = this.config.waveColor?.toString?.() ?? "#000";
+    layer.strokeStyle = waveColor;
+    layer.fillStyle = waveColor;
+    layer.lineWidth = 1;
+    layer.beginPath();
+    layer.moveTo(x, y);
+    const now = performance.now();
+    for (let i = 0; i < bufferChunkSize; i++) {
+      const slice = bufferChunks[i];
+      const sliceLength = slice.length;
+      const chunkStart = Math.floor(clamp(iStart - total, 0, sliceLength));
+      const chunkEnd = Math.ceil(clamp(iEnd - total, 0, sliceLength));
+      total += sliceLength;
+      try {
+        const chunks = slice.slice(chunkStart, chunkEnd);
+        const l = chunks.length - 1;
+        let i = l + 1;
+        while (i > 0) {
+          const index = l - i;
+          const chunk = chunks.slice(index, index + context.samplesPerPx);
+          if (now - performance.now() > 10) {
```

Comment:
I'm fairly certain this would have never yielded previously since the values would have been negative.
```suggestion
          if (performance.now() - now > 10) {
```

Looking at it further, the value of now should be reset after it yields to performance.now() otherwise it won't ever yield again if needed.